sdk/couchbase-core/src/queryx/query_options.rs

```diff
@@ -0,0 +1,238 @@
+#[non_exhaustive]
+pub enum ProfileMode {
+    Off,
+    Phases,
+    Timings,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
+#[non_exhaustive]
+pub enum Compression {
+    Zip,
+    Rle,
+    Lzma,
+    Lzo,
+    None,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
+#[serde(rename_all = "camelCase")]
+#[non_exhaustive]
+pub enum DurabilityLevel {
+    None,
+    Majority,
+    MajorityAndPersistActive,
+    PersistToMajority,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
+#[non_exhaustive]
+pub enum Encoding {
+    #[serde(rename = "UTF-8")]
+    Utf8,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
+#[non_exhaustive]
+pub enum Format {
+    Json,
+    Xml,
+    Csv,
+    Tsv,
+}
+
+#[derive(Debug, Clone, Serialize)]
+pub struct CredsJson {
+    #[serde(skip_serializing_if = "String::is_empty")]
+    pub user: String,
+    #[serde(skip_serializing_if = "String::is_empty")]
+    pub pass: String,
+}
+
+#[derive(Debug, Clone, Serialize)]
+pub struct ScanVectorEntry {
+    pub seq_no: u64,
+    pub vb_uuid: String,
+}
+
+#[derive(Debug, Clone, Serialize)]
+#[serde(rename_all = "lowercase")]
+#[non_exhaustive]
+pub enum ReplicaLevel {
+    Off,
+    On,
+}
+
+pub type FullScanVectors = Vec<ScanVectorEntry>;
+pub type SparseScanVectors = HashMap<u32, ScanVectorEntry>;
+
+#[derive(Debug, Clone, TypedBuilder)]
+#[builder(field_defaults(setter(into)))]
+#[non_exhaustive]
+pub struct QueryOptions {
+    pub args: Vec<Vec<u8>>,
+    pub atr_collection: Option<String>,
+    pub auto_execute: Option<bool>,
+    pub client_context_id: Option<String>,
+    pub compression: Option<Compression>,
+    pub controls: Option<bool>,
+    pub creds: Option<Vec<CredsJson>>,
+    pub durability_level: Option<DurabilityLevel>,
+    pub encoded_plan: Option<String>,
+    pub encoding: Option<Encoding>,
+    pub format: Option<Format>,
+    pub kv_timeout: Option<Duration>,
+    pub max_parallelism: Option<u32>,
+    pub memory_quota: Option<u32>,
+    pub metrics: Option<bool>,
+    pub namespace: Option<String>,
+    pub num_atrs: Option<u32>,
+    pub pipeline_batch: Option<u32>,
+    pub pipeline_cap: Option<u32>,
+    pub prepared: Option<String>,
+    pub preserve_expiry: Option<bool>,
+    pub pretty: Option<bool>,
+    pub profile: Option<ProfileMode>,
+    pub query_context: Option<String>,
+    pub read_only: Option<bool>,
+    pub scan_cap: Option<u32>,
+    pub scan_consistency: Option<ScanConsistency>,
+    pub scan_vector: Vec<u8>,
+    pub scan_vectors: HashMap<String, Vec<u8>>,
+    pub scan_wait: Option<Duration>,
+    pub signature: Option<bool>,
+    pub statement: Option<String>,
+    pub timeout: Option<Duration>,
+    pub tx_data: Vec<u8>,
+    pub tx_id: Option<String>,
+    pub tx_implicit: Option<bool>,
+    pub tx_stmt_num: Option<u32>,
+    pub tx_timeout: Option<Duration>,
+    pub use_cbo: Option<bool>,
+    pub use_fts: Option<bool>,
+    pub use_replica: Option<ReplicaLevel>,
```

Comment:
I suppose this is an enum as opposed to `Option<bool>` for easier serialisation?

Reply:
I don't think it's bool on the wire and I can envision a world where query adds a new option in the future.